### PR TITLE
Introduce cache / memory methods for costMatrix

### PR DESCRIPTION
**Pull request by TooAngel**
CostMatrixes are stored in plain format in the cache.
CostMatrixes of controlled rooms are stored in `Memory` serialized.

### CODE REVIEW

**Comment by TooAngel** [comment]
DCBOT: 
This repository is under [democratic collaboration](https://github.com/TooAngel/democratic-collaboration) and will be merged automatically.

The merge decision is based on the outcome of the reviews:
 - `Approve` add the reviewer value (number of commits) to the `metric`
 - `Request changes` substract the reviewer value from the `metric`

 The merge will happen after `metric * 10 days` calculated from the last code change.

Please review the PR to make a good democratic decision.

Summoning some reviewers:

 - @Tsungen: Some warm words would be nice - @Evanito: Some warm words would be nice - @mschultheiss83: I always like your reviews

**Comment by TooAngel** [comment]
DCBOT: Code update recognized, countdown starts fresh.

**Comment by TooAngel** [comment]
DCBOT: Code update recognized, countdown starts fresh.

File: src/config_brain_memory.js [hunk 1]
```diff
@@ -47,7 +47,10 @@ brain.prepareMemory = function() {
   for (let name in Memory.creeps) {
     if (!Game.creeps[name]) {
       let role = Memory.creeps[name].role;
-      let roleStat = Memory.stats[userName].roles[role];
+      let roleStat;
+      if (!Memory.stats[userName]) {
+        roleStat = Memory.stats[userName].roles[role];
+      }
```
Comment (Nevrdid) [inline src/config_brain_memory.js]: not sure this is correct ^^'

**Comment by TooAngel** [comment]
DCBOT: A new review, yeah.

    Votes: 99/285
    Coefficient: 0.347368421053
    Days to merge: 6.52631578947
    Age: 0

**Comment by TooAngel** [comment]
DCBOT: Code update recognized, countdown starts fresh.

**Comment by TooAngel** [comment]
DCBOT: Adding WIP label, the title is prefixed with [WIP]

File: src/prototype_creep_routing.js [hunk 2]
```diff
@@ -15,38 +15,9 @@ Creep.prototype.getRoute = function() {
...
-    };
-    route = Game.map.findRoute(this.memory.base, this.memory.routing.targetRoom, {
-      routeCallback: routeCallback
-    });
-    //     this.log('getRoute: ' + this.memory.base + ' target: ' + this.memory.routing.targetRoom + ' route: ' + route);
-    // TODO I guess some bug while transitionen to memory.routing
-    if (route == -2) {
-      route = [];
-    }
+    route = this.room.findRoute(this.memory.base, this.memory.routing.targetRoom);
```
Comment (TooAngel) [inline src/prototype_creep_routing.js]: We always calculate the routes from `base` to `targetRoom`. In an older version I used `this.room.name`, but I ran into different kind of problems.

If we have cases where we want to have it different, we can discuss. For now I would keep it as it is.

**Comment by TooAngel** [comment]
DCBOT: Code update recognized, countdown starts fresh.

**Comment by TooAngel** [comment]
DCBOT: Removing WIP label, the title is not prefixed with [WIP]

**Comment by TooAngel** [comment]
DCBOT: Adding WIP label, the title is prefixed with [WIP]

**Comment by TooAngel** [comment]
DCBOT: Adding WIP label, the title is prefixed with [WIP]

**Comment by Somotaw** [comment]
I like the carry dismantle code, working like a charm when cherry picked the commit ;-*

**Comment by Somotaw** [comment]
woops nvm its dismantling containers ;)

**Comment by TooAngel** [comment]
DCBOT: Code update recognized, countdown starts fresh.

**Comment by TooAngel** [comment]
DCBOT: Code update recognized, countdown starts fresh.

**Comment by TooAngel** [comment]
DCBOT: Code update recognized, countdown starts fresh.

**Comment by TooAngel** [comment]
DCBOT: Code update recognized, countdown starts fresh.

**Comment by TooAngel** [comment]
DCBOT: Code update recognized, countdown starts fresh.

**Comment by TooAngel** [comment]
DCBOT: Code update recognized, countdown starts fresh.

**Comment by TooAngel** [comment]
DCBOT: Code update recognized, countdown starts fresh.

**Comment by TooAngel** [comment]
DCBOT: Removing WIP label, the title is not prefixed with [WIP]

File: src/config.js [hunk 3]
```diff
@@ -29,7 +29,7 @@ global.config = {
 
   info: {
     signController: true,
-    signText: 'Fully automated TooAngel bot: https://github.com/TooAngel/screeps',
+    signText: 'Fully automated TooAngel bot: http://tooangel.github.io/screeps/',
```
Comment (Somotaw) [inline src/config.js]: Nice :)

**Comment by Somotaw** [comment]
Testing this now.

**Comment by Somotaw** [comment]
so far [12:04:59]16890238 W31N76 scout-180            [room W31N76 pos 23,0] search: [room W32N76 pos 25,25] TypeError: Cannot read property 'find' of undefined

**Comment by Somotaw** [comment]
seems to have fixed itself with time

**Comment by Somotaw** [comment]
hmm every time I deploy a get a few of these
[12:12:54]16890366 W31N76                      getCostMatrixCallback updatePosition
[12:12:54]16890366 W31N76                      Update position
[12:12:54]16890366 W31N76 scout-180            [room W31N76 pos 23,0] search: [room W32N76 pos 25,25] TypeError: Cannot read property 'find' of undefined
[12:12:56]16890367 W31N77                      getCostMatrixCallback updatePosition
[12:12:56]16890367 W31N77                      Update position
[12:12:56]16890367 W31N77 scout-180            [room W31N77 pos 23,49] search: [room W32N76 pos 25,25] TypeError: Cannot read property 'find' of undefined
[12:13:00]16890368 W31N76                      getCostMatrixCallback updatePosition
[12:13:00]16890368 W31N76                      Update position
[12:13:00]16890368 W31N76 scout-180            [room W31N76 pos 23,0] search: [room W32N76 pos 25,25] TypeError: Cannot read property 'find' of undefined

**Comment by TooAngel** [comment]
DCBOT: Adding WIP label, the title is prefixed with [WIP]

**Comment by Somotaw** [comment]
04:16:52]16905291 W30N75 scout-815            [room W30N75 pos 49,26] search: [room W32N76 pos 25,25] TypeError: Cannot read property 'base' of undefined TypeError: Cannot read property 'base' of undefined
    at Room.getMemoryCostMatrix (prototype_room_memory:46:102)
    at Room.getCostMatrixCallback (prototype_room_costmatrix:4:25)
    at breadthFirstSearch (role_scout:96:36)
    at Object.roles.scout.execute (role_scout:147:10)
    at Creep.handle (prototype_creep:64:12)
    at Room.execute (prototype_room:15:13)
    at Memory.myRooms._.map._.filter.r (main:59:54)
    at /opt/engine/node_modules/lodash/index.js:1932:13
    at /opt/engine/node_modules/lodash/index.js:3073:15
    at baseForOwn (/opt/engine/node_modules/lodash/index.js:2046:14)

**Comment by Somotaw** [comment]
16908773 W30N78 scout-254            [room W30N78 pos 49,20] search: [room W29N79 pos 25,25] TypeError: Cannot read property 'base' of undefined TypeError: Cannot read property 'base' of undefined
    at Room.getMemoryCostMatrix (prototype_room_memory:46:102)
    at Room.getCostMatrixCallback (prototype_room_costmatrix:4:25)
    at breadthFirstSearch (role_scout:96:36)
    at Object.roles.scout.execute (role_scout:147:10)
    at Creep.handle (prototype_creep:64:12)
    at Room.execute (prototype_room:15:13)
    at Memory.myRooms._.map._.filter.r (main:59:54)
    at /opt/engine/node_modules/lodash/index.js:1932:13
    at /opt/engine/node_modules/lodash/index.js:3073:15
    at baseForOwn (/opt/engine/node_modules/lodash/index.js:2046:14)
[07:59:52]16908775 W30N78 scout-254            [room W30N78 pos 49,20] search: [room W29N79 pos 25,25] TypeError: Cannot read property 'base' of undefined TypeError: Cannot read property 'base' of undefined
    at Room.getMemoryCostMatrix (prototype_room_memory:46:102)
    at Room.getCostMatrixCallback (prototype_room_costmatrix:4:25)
    at breadthFirstSearch (role_scout:96:36)
    at Object.roles.scout.execute (role_scout:147:10)
    at Creep.handle (prototype_creep:64:12)
    at Room.execute (prototype_room:15:13)
    at Memory.myRooms._.map._.filter.r (main:59:54)
    at /opt/engine/node_modules/lodash/index.js:1932:13
    at /opt/engine/node_modules/lodash/index.js:3073:15
    at baseForOwn (/opt/engine/node_modules/lodash/index.js:2046:14)

**Comment by Somotaw** [comment]
scout callback is doing something it shouldn,t Have to revert to old scout callback function for it to work for now, without messing with your other stuff.

File: src/config.js (hunk 3 above)
Comment (mschultheiss83) [inline src/config.js]: on http://tooangel.github.io/screeps/ the last line are not well formated (maybe missing newlines)

Comment (mschultheiss83) [inline src/config.js]: but i like

Comment (TooAngel) [inline src/config.js]: The page is created from the README.md. The documentation gets improved with #197 . Should hopefully also fix these issues.

**Comment by democratic-collaboration** [comment]
DCBOT: Removing WIP label, the title is not prefixed with [WIP]

**Pull request merged**